econml/validate/drtester.py

```diff
@@ -0,0 +1,796 @@
+            implemented (with Ztrain specified)
+        n_groups: integer, default 4
+            Number of quantile-based groups used to calculate calibration score.
+
+        Returns
+        -------
+        self, with added attribute cal_r_squared showing the R^2 value of the calibration test and dataframe df_plot
+        containing relevant results to plot calibration test gates
+
+        """
+        if self.dr_train is None:
+            raise Exception("Must fit nuisance models on training sample data to use calibration test")
+
+        # if CATE is given explicitly or has not been fitted at all previously, fit it now
+        if (self.cate_preds_train is None) or (self.cate_preds_val is None) or (reg_cate is not None):
+            if (Zval is None) or (Ztrain is None) or (reg_cate is None):
+                raise Exception('CATE not fitted on training data - must provide CATE model and both Zval, Ztrain')
+            self.fit_cate(reg_cate, Zval, Ztrain)
+
+        self.cal_r_squared = np.zeros(self.n_treat)
+        self.df_plot = pd.DataFrame()
+        for k in range(self.n_treat):
+
+            cuts = np.quantile(self.cate_preds_train[:, k], np.linspace(0, 1, n_groups + 1))
+            probs = np.zeros(n_groups)
+            g_cate = np.zeros(n_groups)
+            se_g_cate = np.zeros(n_groups)
+            gate = np.zeros(n_groups)
+            se_gate = np.zeros(n_groups)
+            for i in range(n_groups):
+                # Assign units in validation set to groups
+                ind = (self.cate_preds_val[:, k] >= cuts[i]) & (self.cate_preds_val[:, k] <= cuts[i + 1])
+                # Proportion of validations set in group
+                probs[i] = np.mean(ind)
+                # Group average treatment effect (GATE) -- average of DR outcomes in group
+                gate[i] = np.mean(self.dr_val[ind, k])
+                se_gate[i] = np.std(self.dr_val[ind, k]) / np.sqrt(np.sum(ind))
+                # Average of CATE predictions in group
+                g_cate[i] = np.mean(self.cate_preds_val[ind, k])
+                se_g_cate[i] = np.std(self.cate_preds_val[ind, k]) / np.sqrt(np.sum(ind))
+
+            # Calculate group calibration score
+            cal_score_g = np.sum(abs(gate - g_cate) * probs)
+            # Calculate overall calibration score
+            cal_score_o = np.sum(abs(gate - self.ate_val[k]) * probs)
+            # Calculate R-square calibration score
+            self.cal_r_squared[k] = 1 - (cal_score_g / cal_score_o)
+
+            df_plot1 = pd.DataFrame({'ind': np.array(range(n_groups)),
+                                     'gate': gate, 'se_gate': se_gate,
+                                    'g_cate': g_cate, 'se_g_cate': se_g_cate})
+            df_plot1['tmt'] = self.tmts[k + 1]
+            self.df_plot = pd.concat((self.df_plot, df_plot1))
+
+        return self
```

Comment:
Most prob this should be returning some evaluation result e.g. the calibration score

Reply:
Maybe you should create a class CalibrationEvaluationResults that contains all these variables you describe and has the API for parsing it, e.g. the plot_cal, most prob should be part of this new class of `CalibrationEvaluationResults`